### ISSUE TITLE
Custom Formatter for airlift Log-manager used by presto

### ISSUE THREAD

**Issue by ssanthanam185** [airlift/airlift#768]
For ingesting logs into Kibana , would like presto logs to be spit out in a JSON format. User should be able to extend java.util.logging.Formatter with a custom implementation & configure Airlift to use this custom class instead of the StaticFormatter.

**Comment by electrum** [comment]
Do you have a specific format in mind? We’d prefer to have an option to logging in a standard JSON format rather than making it completely configurable.

**Comment by ssanthanam185** [comment]
Yes for JSON format. How ever Id like the flexibility in mapping some fields in the record to standard fields which have been defined in Kibana and have indexes built on it.

**Comment by electrum** [comment]
Can you give a specific example of what you’d like a log message to look like?

**Comment by ssanthanam185** [comment]
Example : 
{ "ts" ="2019-10-29T23:47:40.63127", 
"lvlname" ="WARN",
"name":"io.prestosql.server.remotetask.RequestErrorTracker",
"exc":"Error updating task 20191029_234709_13560_v23z9.21.22: java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed: http://10.46.223.209:8081/v1/task/20191029_234709_13560_v23z9.21.22"
}

Would be nice if users can add their custom Formatter & just define it as a config like: 

log.custom-formatter=org.formmater.JsonFormatter 

We should be able to add org.formmater.JsonFormatter classpath, so that Airlift can load it and inject it as the Formatter to be used in the OutputstreamHandler

**Comment by cccs-tom** [comment]
We have a similar stack and would very much appreciate this feature as well.

**Comment by electrum** [comment]
We don't use SLF4J in Trino because of class loaders. If plugins log via JUL, then everything just works. If the use SLF4J, then the plugin has to bridge to JUL, or we would need to add SLF4J to the plugin SPI, and that's problematic for a variety of reasons.

We recently added a JSON logging format:

```
log.format=json
```

There is an optional annotations file which will be included as an `annotations` map in the JSON log record:

```
log.annotation-file=/path/to/annotations.properties
```

Finally, you can use a TCP destination for logging:

```
log.path=tcp://host:port
```

Note that both the logging configuration properties and the annotations file properties support the environment variable replacement that is described here: https://trino.io/docs/current/security/secrets.html

**Comment by lfrancke** [comment]
Thanks for the detailed response.
That helps and I do understand the reasoning. It is a trade-off for sure.

Apache NiFi is also based on a plugin architecture and they made a different [decision](https://nifi.apache.org/developer-guide.html). They have a `ComponentLog` interface which can then be implemented in various ways.

As I said: I do understand your decision, it's a valid one.
Would you be up to reconsider that decision in any way or is it moot discussing the point (at the moment)?

**Comment by electrum** [comment]
Thanks for the pointer. We considered adding our own logging interface (and already use Airlift [Logger](https://github.com/airlift/airlift/blob/master/log/src/main/java/io/airlift/log/Logger.java) directly in Trino), but using JUL didn't require changing existing code or writing new logging bridges.

Making logging completely pluggable has never been a goal for Trino, so I'd prefer to focus on the underlying problem you're trying to solve. What's your goal with customized logging?

If you change the Trino `Server` class to call `doNotInitializeLogging()` on `Bootstrap()`, then it won't touch the JUL configuration, so you could have a custom JUL handler or JUL-to-SLF4J bridge. I think that should fix the issues you found in your attempt, but there might be another problem I'm not seeing.

**Comment by lfrancke** [comment]
Our goal is mostly consistency. All other Java tools we use (HBase, NiFi, Spark, Kafka....) are easily configurable to emit the exact same logging format. Trino is the only outlier which requires special handling as it uses "non-standard" logging. I put it into quotes as I do realize that you're using the only logging implementation that actually comes out of the box ;-)

It's not a huge deal, we just found this issue and thought we'd inquire.
We'd be willing to do the work as well but as I said: It only makes sense if you'd accept any changes here which doesn't seem to be the case and that's fine.

We did play around with JUL-to-SLF4J but that wouldn't work 100%. It would always emit a "header" anyway (we didn't dig too deep) and then our properly formatted log message - and that again would require special handling. Now we're using the JSON format instead.

Thank you for your answers, that already helps us. There is nothing that needs to be done here as far as we're concerned. I would suggest to leave this issue open to collect feedback from others.

**Comment by deigote** [comment]
> It would always emit a "header" anyway (we didn't dig too deep) and then our properly formatted log message - and that again would require special handling. Now we're using the JSON format instead.

maybe it's something else in your case, but I had this issue with "special header + my message" too, which broke our logs. I found out that [airlift does a global stdin redirect](https://github.com/airlift/airlift/blob/a42258ddd7b2317ddc1dbbde1d7e0dd0cb73312a/log-manager/src/main/java/io/airlift/log/Logging.java#L139) (which was a very big WTF moment for me, TBH... maybe there are good reasons), so that all output is processed by the logging system (so then something like `System.out.println("some info")` looks like a log). I'm right now undoing this in my plugins with `System.setOut(new PrintStream(new FileOutputStream(FileDescriptor.out)));` which is... oof. But not sure how else to go about it (I need my plugins to log as JSON and with lots of metadata, so I need JSON and MDC which airlift doesn't seem to support properly yet).

**Comment by lihonosov** [comment]
I would like to contribute my thoughts on this matter. It is commendable that Trino supports `log.format=JSON` in addition to `log.path=tcp|file`. However, deploying Trino to Kubernetes and integrating it with monitoring systems like Datadog involves some complex steps to configure the logs correctly. By default, Datadog retrieves logs from stdin, and logs formatted in JSON are displayed effectively within Datadog. Conversely, logs not in JSON format are often misinterpreted as errors, and stack traces are split across multiple lines. I believe this issue is not unique to Datadog but also affects other monitoring systems such as Prometheus. Nowadays, the support for structured logging is a highly appreciated feature and has become a de facto standard.

https://cloud.google.com/logging/docs/structured-logging
https://www.pluralsight.com/resources/blog/cloud/how-to-debug-faster-with-structured-logging
https://newrelic.com/blog/how-to-relic/structured-logging
https://www.elastic.co/blog/structured-logging-filebeat

@electrum any plans to implement `log.format=JSON` for stdin as well?  Is there a specific reason why this property is effective only when `log.path` is set and not as a global default?

**Comment by electrum** [comment]
@mattstep Do you remember why we don't support JSON for stdout (console)? My guess is that the original use case was to have JSON logs for Datadog or other systems with stdout for humans (using `kubectl logs` or similar). I think we should support JSON for stdout, but might want a flag to allow the current behavior.

**Comment by sgoncharov1** [comment]
Hi, 
First of all, thank you for all the work you are investing in the Trino product ! 

We also have a similar use case, and would be happy for an option of printing logs in JSON format to stdout .
(similar use case to what @lihonosov mentioned on Jun 28 ) 
This will allow us to simplify the deployment architecture.

**Comment by megri** [comment]
Currently having a hard time trying to get Trino (which by proxy means Airlift) to output logs structured to GCP standards.

All of this would be a non-issue if Airlift didn't clear out all user-land logging configuration, rewrite stderr/stdout on a global level and force `StaticFormatter` on everything. Why not let users configure the logging output if so desired? I don't understand..
